pkg/datastore/deploymentchainstore.go

```diff
@@ -59,22 +61,74 @@ var (
 			if blockIndex >= uint32(len(dc.Blocks)) {
 				return fmt.Errorf("invalid block index %d provided", blockIndex)
 			}
+
+			var (
+				updated                bool
+				successDeploymentCtn   int
+				failedDeploymentCtn    int
+				cancelledDeploymentCtn int
+				runningDeploymentCtn   int
+			)
+
 			block := dc.Blocks[blockIndex]
-			var updated bool
 			for _, node := range block.Nodes {
 				if node.DeploymentRef == nil {
 					continue
 				}
-				if node.DeploymentRef.DeploymentId != deploymentID {
-					continue
+				if node.DeploymentRef.DeploymentId == deploymentID {
+					node.DeploymentRef.Status = status
+					node.DeploymentRef.StatusReason = reason
+					updated = true
+				}
+
+				// Count values to determine block status.
+				switch node.DeploymentRef.Status {
+				case model.DeploymentStatus_DEPLOYMENT_SUCCESS:
+					successDeploymentCtn++
+				case model.DeploymentStatus_DEPLOYMENT_FAILURE:
+					failedDeploymentCtn++
+				case model.DeploymentStatus_DEPLOYMENT_CANCELLED:
+					cancelledDeploymentCtn++
+				case model.DeploymentStatus_DEPLOYMENT_RUNNING:
+					runningDeploymentCtn++
 				}
-				node.DeploymentRef.Status = status
-				node.DeploymentRef.StatusReason = reason
-				updated = true
 			}
+
 			if !updated {
 				return fmt.Errorf("unable to find the right node in chain to assign deployment to")
 			}
+
+			// If the block is already finished, keep its finished status.
+			if block.IsCompleted() {
+				return nil
+			}
+
+			// Update block status based on its state after update latest submitted deployment status.
+			// If all the nodes in block is completed successfully, the block counted as SUCCESS.
+			if successDeploymentCtn == len(block.Nodes) {
+				block.Status = model.ChainBlockStatus_DEPLOYMENT_BLOCK_SUCCESS
+				block.CompletedAt = time.Now().Unix()
+				return nil
+			}
+			// If one of the node in the block is completed with FAILURE status, the block counted as FAILURE.
+			if failedDeploymentCtn > 0 {
+				block.Status = model.ChainBlockStatus_DEPLOYMENT_BLOCK_FAILURE
+				block.CompletedAt = time.Now().Unix()
+				return nil
+			}
+			// If one of the node in the block is completed with CANCELLED status, the block counted as CANCELLED.
+			if cancelledDeploymentCtn > 0 {
+				block.Status = model.ChainBlockStatus_DEPLOYMENT_BLOCK_CANCELLED
+				block.CompletedAt = time.Now().Unix()
+				return nil
+			}
+			// If there is at least a deployment in chain which has RUNNING status,
+			// and the block passed all above filters, the block counted as RUNNING.
+			if runningDeploymentCtn > 0 {
+				block.Status = model.ChainBlockStatus_DEPLOYMENT_BLOCK_RUNNING
+			}
+			// Otherwise, the block status is remained.
+
```

Comment:
How about moving this kind of logic to the model package?

Reply:
Another node iteration is added but it looks more clear than it is. Nice point, lets me adopt that. (tbh, I don't remember why I placed this status determining block here 🤔 )

Reply:
Looks definitely better 👍 PTAL https://github.com/pipe-cd/pipe/pull/2866/commits/b4f162133aaf39ee955c04a10ad8f16bd8edc3e6